source/includes/fundamentals/code-snippets/GridFSOperations.java

```diff
@@ -0,0 +1,168 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+import com.mongodb.MongoGridFSException;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.gridfs.GridFSBucket;
+import com.mongodb.client.gridfs.GridFSBuckets;
+import com.mongodb.client.gridfs.GridFSDownloadStream;
+import com.mongodb.client.gridfs.GridFSUploadStream;
+import com.mongodb.client.gridfs.model.GridFSDownloadOptions;
+import com.mongodb.client.gridfs.model.GridFSFile;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Sorts;
+
+public final class GridFSOperations {
+
+    private static void createCustomBucket(MongoClient mongoClient) throws Exception {
+        MongoDatabase database = mongoClient.getDatabase("mydb");
+        // start createCustomGridFSBucket
+        GridFSBucket gridFSBucket = GridFSBuckets.create(database, "myCustomBucket");
+        // end createCustomGridFSBucket
+    }
+
+    private static void uploadOptions() {
+        // start uploadOptions
+        GridFSUploadOptions options = new GridFSUploadOptions()
+                .chunkSizeBytes(1048576) // 1MB chunk size
+                .metadata(new Document("myField", "myValue"));
+        // end uploadOptions
+    }
+
+    private static void uploadFromInputStream(GridFSBucket gridFSBucket) throws Exception {
+        // start uploadFromInputStream
+        String filePath = "/path/to/project.zip";
+        try (InputStream streamToUploadFrom = new FileInputStream(filePath) ) {
+            GridFSUploadOptions options = new GridFSUploadOptions()
+                    .chunkSizeBytes(1048576)
+                    .metadata(new Document("type", "zip archive"));
+
+            ObjectId fileId = gridFSBucket.uploadFromStream("myProject.zip", streamToUploadFrom, options);
+
+            System.out.println("The file id of the uploaded file is: " + fileId.toHexString());
+        }
+        // end uploadFromInputStream
+    }
+
+    private static void uploadFromOutputStream(GridFSBucket gridFSBucket) throws Exception {
+        // start uploadFromOutputStream
+        Path filePath = Paths.get("/path/to/project.zip");
+        byte[] data = Files.readAllBytes(filePath);
+
+        GridFSUploadOptions options = new GridFSUploadOptions()
+                .chunkSizeBytes(1048576)
+                .metadata(new Document("type", "zip archive"));
+
+        try (GridFSUploadStream uploadStream = gridFSBucket.openUploadStream("myProject.zip", options)) {
+            try {
+                uploadStream.write(data);
+                System.out.println("The file id of the uploaded file is: " + uploadStream.getObjectId().toHexString());
+            } catch (MongoGridFSException e) {
```

Comment:
I think this won't work in the way you'd expect. 

The exception would come from  an insertOne and wouldn't be wrapped in a MongoGridFSException.  The MongoGridFSException would only happen if  `uploadStream.close()` had been called and then the abort call wouldn't work.

Reply:
Catching Exception or MongoException maybe better. But if the insert failed there is a chance calling abort() will also fail.

Reply:
Will remove per our discussion on Slack.